File: server/src/models/SongRequest.js
/**
 * SongRequest Model - Request lagu dari user
 * File: src/models/SongRequest.js
 */

const { DataTypes } = require('sequelize');
const { sequelize } = require('../config/database');

const SongRequest = sequelize.define('SongRequest', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  user_id: {
    type: DataTypes.INTEGER,
    allowNull: false
  },
  title: {
    type: DataTypes.STRING(255),
    allowNull: false
  },
  artist: {
    type: DataTypes.STRING(255),
    allowNull: false
  },
  genre: {
    type: DataTypes.STRING(100),
    allowNull: true
  },
  language: {
    type: DataTypes.STRING(50),
    allowNull: true
  },
  link: {
    type: DataTypes.STRING(500),
    allowNull: true,
    comment: 'Link YouTube/Spotify'
  },
  notes: {
    type: DataTypes.TEXT,
    allowNull: true
  },
  status: {
    type: DataTypes.ENUM('pending', 'approved', 'rejected', 'completed'),
    defaultValue: 'pending'
  },
  admin_notes: {
    type: DataTypes.TEXT,
    allowNull: true,
    comment: 'Catatan dari admin'
  },
  reviewed_by: {
    type: DataTypes.INTEGER,
    allowNull: true
  },
  reviewed_at: {
    type: DataTypes.DATE,
    allowNull: true
  }
}, {
  tableName: 'song_requests',
  timestamps: true,
  createdAt: 'created_at',
  updatedAt: 'updated_at'
});

module.exports = SongRequest;

File: native/lib/features/requests/models/song_request.dart
import 'package:equatable/equatable.dart';

class SongRequest extends Equatable {
  final int id;
  final String title;
  final String artist;
  final String? genre;
  final String? language;
  final String? link;
  final String status;
  final String? notes;

  const SongRequest({
    required this.id,
    required this.title,
    required this.artist,
    this.genre,
    this.language,
    this.link,
    required this.status,
    this.notes,
  });

  factory SongRequest.fromJson(Map<String, dynamic> json) {
    return SongRequest(
      id: _parseInt(json['id']),
      title: json['title']?.toString() ?? '',
      artist: json['artist']?.toString() ?? '',
      genre: json['genre']?.toString(),
      language: json['language']?.toString(),
      link: json['link']?.toString(),
      status: json['status']?.toString() ?? 'pending',
      notes: json['notes']?.toString(),
    );
  }

  @override
  List<Object?> get props => [id, title, artist, genre, language, link, status, notes];

  static int _parseInt(dynamic value) {
    if (value == null) return 0;
    if (value is int) return value;
    return int.tryParse(value.toString()) ?? 0;
  }
}
